bluegreen_deploy.py

```diff
@@ -65,226 +66,305 @@ def list_marathon_apps(args):
+       Return True if we catch an exception while making a request to
+       haproxy or if more than one pid is returned
+    """
+    try:
+        pids = fetch_haproxy_pids(haproxy_url)
+    except requests.exceptions.RequestException as e:
+        return True
+
+    if len(pids) > 1:
+        logger.info("Waiting for {} pids on {}".format(len(pids), haproxy_url))
+        return True
+
+    return False
+
+
+def any_marathon_lb_reloading(marathon_lb_urls):
+    return any([check_haproxy_reloading(url) for url in marathon_lb_urls])
+
+
+def fetch_haproxy_stats(haproxy_url):
+    try:
+        response = requests.get(haproxy_url + "/haproxy?stats;csv")
+        response.raise_for_status()
+    except requests.exceptions.RequestException as e:
+        logger.exception("Caught exception when retrieving HAProxy"
+                         " stats from " + haproxy_url)
+        raise e
+
+    return response.text
+
+
+def fetch_combined_haproxy_stats(marathon_lb_urls):
+    raw = ''.join([ fetch_haproxy_stats(url) for url in marathon_lb_urls])
+    return parse_haproxy_stats(raw)
+
+
+def parse_haproxy_stats(csv_data):
+    rows = csv_data.splitlines()
+    headings = rows.pop(0).lstrip('# ').rstrip(',\n').split(',')
+    csv_reader = csv.reader(rows, delimiter=',', quotechar="'")
+
+    Row = namedtuple('Row', headings)
+
+    return [Row(*row[0:-1]) for row in csv_reader if row[0][0] != '#']
+
+
+def get_deployment_label(app):
+    return get_deployment_group(app) + "_" + app['labels']['HAPROXY_0_PORT']
+
+
+def _if_app_listener(app, listener):
+    return listener.pxname == get_deployment_label(app) and \
+      listener.svname not in ['BACKEND', 'FRONTEND']
+
+
+def fetch_app_listeners(app, marathon_lb_urls):
+    haproxy_stats = fetch_combined_haproxy_stats(marathon_lb_urls)
+    return [l for l in haproxy_stats if _if_app_listener(app, l)]
+
+
+def waiting_for_listeners(new_app, old_app, listeners, haproxy_count):
+    listener_count = (len(listeners) / haproxy_count)
+    return listener_count != new_app['instances'] + old_app['instances']
+
+
+def get_deployment_target(app):
+    return int(app['labels']['HAPROXY_DEPLOYMENT_TARGET_INSTANCES'])
+
+
+def waiting_for_up_listeners(app, listeners, haproxy_count):
+    up_listeners = [l for l in listeners if l.status == 'UP']
+    up_listener_count = (len(up_listeners) / haproxy_count)
+
+    return up_listener_count < get_deployment_target(app)
+
+
+def _has_pending_requests(listener):
+    return int(listener.qcur or 0) > 0 or int(listener.scur or 0) > 0
+
+
+def select_drained_listeners(listeners):
+    draining_listeners = [l for l in listeners if l.status == 'MAINT']
+    return [l for l in draining_listeners if not _has_pending_requests(l)]
+
+
+def get_svnames_from_task(task):
+    prefix = '{}_{}_{}_{}'.format(*task['host'].split('.'))
+    for port in task['ports']:
+        yield(prefix + '_{}'.format(port))
+
+
+def get_svnames_from_tasks(tasks):
+    svnames = []
```

Comment:
Problem get_svnames_from_task returns lists... i'll have to figure out how to fix that before i can use the list comp here